src/codegen/compile-type-definition.lisp

```diff
@@ -54,33 +75,25 @@
                                      (values ,classname))
                             (format stream "#.(~s~{ ~s~})"
                                     ',(constructor-entry-name constructor)
-                                    (list ,@(mapcar (lambda (slot) `(slot-value self ',slot)) slot-names)))
+                                    (list ,@(mapcar (lambda (slot) `(slot-value self ',slot)) field-names)))
                             self)))
 
              :append (cond
-                       ((= 0 (constructor-entry-arity constructor))
-                        `((coalton-impl::define-global-lexical
-                              ,(constructor-entry-name constructor)
-                              (,(constructor-entry-name constructor)))))
+                       ((zerop (constructor-entry-arity constructor))
+                        (list `(coalton-impl::define-global-lexical
+                                   ,(constructor-entry-name constructor)
+                                   (,(constructor-entry-name constructor)))))
                        (t
-                        (let* ((arity (length slot-names))
+                        (let* ((arity (length field-names))
                                (entry (construct-function-entry
                                        `#',(constructor-entry-name constructor)
                                        arity)))
-                          `((coalton-impl::define-global-lexical
-                                ,(constructor-entry-name constructor)
-                              ,entry))))))
-         #+sbcl
-         (declaim (sb-ext:freeze-type ,(type-definition-name def))))))))
-
-(defun ctor-make-slot-names (count package)
-  (loop :for i :below count
-        :collect (alexandria:format-symbol package "_~D" i)))
-
-(defun ctor-make-slot (name type env)
-  `(,name (error "") :type ,(lisp-type type env) :read-only t))
+                          (list `(coalton-impl::define-global-lexical
+                                     ,(constructor-entry-name constructor)
+                                   ,entry))))))
 
-(defun ctor-make-slots (slot-names slot-types env)
-  (loop :for slot-name :in slot-names
-        :for slot-type :in slot-types
-        :collect (ctor-make-slot slot-name slot-type env)))
+         ,@(when (eql coalton-impl::*interaction-mode* :release)
+             #+sbcl
```

Comment:
I think in this case you should splat out a list which could contain multiple possible flags (for non-sbcl impl specific stuff) and just have `#+sbcl` to add to the list